---
title: "- Exploratory Data Analysis -"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: true
    toc_float: true
    number_sections: true
    df_print: paged
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

<style>
body{text-align: justify}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>

*File creation: October, 13th 2021*  
*Update: May, 03rd 2022*  

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(here)
library(Seurat)
library(MAST)
library(doMC)
library(cowplot)
library(data.table)
```
```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
load(here("output", "001_output", "001_EDA.RData"))
# save.image(here("output", "001_output", "001_EDA.RData"))
```

# Exploratory Data Analysis
___________________________

Standard analyses are performed using the `Seurat` R package (Bulter et al., 2019). Here, the goal is to integrate the dataset using sample (combination of each batch, PTID, sorted population, arm and stimulation) as factor using their anchor-based workflow, enabling cells to cluster together based on their shared biological state, as opposed to sample-of-origin. It is important to integrate because the loss of variability between samples is both expected and desirable. The main motivation for performing data correction is to enable us to characterize population heterogeneity in a consistent manner across samples, and to learn cell-type specific responses in downstream analyses.

Quality control has already been done (see `000_Quality_Control.Rmd`). All batch/pools including all samples are merged together, resulting in a dataset of **241,428 cells** (AgSpec+ and AgSpecNot+).  
Gene expression data integration is performed using sample (batch, PTID, arm, stimulation & visit) as factor, resulting in 64 different datasets. To integrate the gene expression values, we first separately normalize each of our partitions using the `SCTransform`. Data integration is then performed using `FindIntegrationAnchors` and `IntegrateData` functions (with `dims = 1:50`, `reduction = rpca` and `reference = 1:2` corresponding to the first PTID & batch). Principal component analysis (PCA) on the integrated dataset is performed using the `RunPCA` R function. The top 50 principal components (PCs) are then used for UMAP visualization and clustering. UMAP with these parameters (`min.dist = .01`) is performed using the `RunUMAP` R function. Cell clustering is performed using a graph-based clustering method implemented in Seurat (`FindNeighbors` and `FindClusters` R function) with different resolutions.  
Data integration using Antibody-derived tags (ADT) is also performed using sample (batch, PTID, arm, stimulation & visit) as factor. For each of our partitions, a centered log ratio transformation (CLR) is performed (`NormalizeData` with `normalization.method = "CLR"`). Data integration and visualization are then performed using the same workflow as described above.  
The weighted nearest neighbor (WNN) workflow is then investigated. This workflow consists of three steps: 1) independent preprocessing and dimensional reduction of each modality individually; 2) learning cell-specific modality *weights*, and constructing a WNN graph that integrates the modalities; and 3) downstream analysis (i.e. visualization, clustering, etc.) of the WNN graph. The `FindMultiModalNeighbors` R function is run with `dims.list = list(1:50, 1:50)` and `modality.weight.name = c("RNA.weight", "ADT.weight")`.  

```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Integration - GEX - using as factor combination of batch, ptid, arm, stimulation & visit
# Stimulation used as proxy for sorted population
merged.seurat <- readRDS(file = here("data", "001_merged-seurat.rds"))
merged.seurat <- subset(merged.seurat, sort_population != "CD3+")
merged.seurat@meta.data$batch <- sapply(X = merged.seurat@meta.data$orig.ident, FUN = function(x) str_split(string = x, pattern = "_")[[1]][1])
merged.seurat@meta.data$sample <- paste(merged.seurat@meta.data$batch, merged.seurat@meta.data$ptid, merged.seurat@meta.data$arm, merged.seurat@meta.data$stim, merged.seurat@meta.data$visit, sep = "_")
DefaultAssay(object = merged.seurat) <- "RNA"
seurat.list <- SplitObject(merged.seurat, split.by = "sample")

#- Normalization
for (i in 1:length(seurat.list)) {
  cat("\nNormalization -", i)
  seurat.list[[i]] <- SCTransform(seurat.list[[i]], verbose = TRUE)
}

#- Data Integration (with features, reference & rpca)
features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 2000, verbose = TRUE)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features, verbose = TRUE)
for (i in 1:length(seurat.list)) {
  cat("\nRunning PCA -", i)
  seurat.list[[i]] <- RunPCA(seurat.list[[i]], features = features, verbose = TRUE)
}
seurat.anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", dims = 1:50, anchor.features = features, reduction = "rpca", reference = 1:2, verbose = TRUE) # One patient is used as reference
# saveRDS(object = seurat.anchors, file = here("output", "001_output", "anchors_GEX.rds"))
integrated.seurat <- IntegrateData(anchorset = seurat.anchors, normalization.method = "SCT", dims = 1:50, verbose = TRUE)
DefaultAssay(integrated.seurat) <- "integrated"

#- Normalization
integrated.seurat <- NormalizeData(object = integrated.seurat, assay = "RNA", normalization.method = "LogNormalize", verbose = TRUE)
integrated.seurat <- NormalizeData(object = integrated.seurat, assay = "ADT", normalization.method = "CLR", verbose = TRUE)

#- PCA
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRAJ"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRBJ"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRGJ"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRDJ"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRAD"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRBD"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRGD"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRDD"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRAV"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRBV"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRGV"))]
VariableFeatures(integrated.seurat) <- VariableFeatures(integrated.seurat)[which(!str_detect(string = VariableFeatures(integrated.seurat), pattern = "^TRDV"))]
integrated.seurat <- RunPCA(object = integrated.seurat, npcs = 100, verbose = TRUE)
ElbowPlot(integrated.seurat, ndims = 100)
chosen.nPCs <- 50

#- UMAP
integrated.seurat <- RunUMAP(object = integrated.seurat, dims = 1:chosen.nPCs, verbose = TRUE, seed.use = 1234, min.dist = .01)

#- Output
saveRDS(object = integrated.seurat, file = here("data", "001_integrated-seurat_GEX.rds"))
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Integration - ADT - using as factor combination of batch, ptid, arm, stimulation & visit
# Stimulation used as proxy for sorted population
merged.seurat <- readRDS(file = here("data", "001_merged-seurat.rds"))
merged.seurat <- subset(merged.seurat, sort_population != "CD3+")
merged.seurat@meta.data$batch <- sapply(X = merged.seurat@meta.data$orig.ident, FUN = function(x) str_split(string = x, pattern = "_")[[1]][1])
merged.seurat@meta.data$sample <- paste(merged.seurat@meta.data$batch, merged.seurat@meta.data$ptid, merged.seurat@meta.data$arm, merged.seurat@meta.data$stim, merged.seurat@meta.data$visit, sep = "_")
DefaultAssay(object = merged.seurat) <- "ADT"
seurat.list <- SplitObject(merged.seurat, split.by = "sample")

#- Normalization
for (i in 1:length(seurat.list)) {
  cat("\nNormalization -", i)
  seurat.list[[i]] <- NormalizeData(seurat.list[[i]], assay = "ADT", normalization.method = "CLR", verbose = TRUE)
  seurat.list[[i]] <- FindVariableFeatures(object = seurat.list[[i]], assay = "ADT")
}

#- Data Integration (with reference & rpca)
features <- SelectIntegrationFeatures(object.list = seurat.list)
for (i in 1:length(seurat.list)) {
  cat("\nScaling & Running PCA -", i)
  seurat.list[[i]] <- ScaleData(object = seurat.list[[i]], assay = "ADT", features = features, vars.to.regress = "nCount_ADT", verbose = TRUE)
  seurat.list[[i]] <- RunPCA(object = seurat.list[[i]], assay = "ADT", features = features, verbose = TRUE)
}
seurat.anchors <- FindIntegrationAnchors(object.list = seurat.list, dims = 1:50, reduction = "rpca", reference = 1:2, verbose = TRUE) # One patient is used as reference
# saveRDS(object = seurat.anchors, file = here("output", "001_output", "anchors_ADT.rds"))
integrated.seurat <- IntegrateData(anchorset = seurat.anchors, dims = 1:50, verbose = TRUE)
DefaultAssay(integrated.seurat) <- "integrated"

#- Normalization
integrated.seurat <- NormalizeData(object = integrated.seurat, assay = "RNA", normalization.method = "LogNormalize", verbose = TRUE)
integrated.seurat <- NormalizeData(object = integrated.seurat, assay = "ADT", normalization.method = "CLR", verbose = TRUE)

#- Scaling & PCA
integrated.seurat <- ScaleData(object = integrated.seurat, vars.to.regress = "nCount_ADT",verbose = TRUE)
integrated.seurat <- RunPCA(object = integrated.seurat, npcs = 100, verbose = TRUE)
ElbowPlot(integrated.seurat, ndims = 100)
chosen.nPCs <- 50

#- UMAP
integrated.seurat <- RunUMAP(object = integrated.seurat, dims = 1:chosen.nPCs, verbose = TRUE, seed.use = 1234, min.dist = .01)

#- Output
saveRDS(object = integrated.seurat, file = here("data", "001_integrated-seurat_ADT.rds"))
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Data Integration using WNN
# remotes::install_github("satijalab/seurat", ref = "release/4.0.0")
integrated.seurat_GEX <- readRDS(file = here("data", "001_integrated-seurat_GEX.rds"))
#ElbowPlot(integrated.seurat_GEX, ndims = 100) # 50 PCs
integrated.seurat_ADT <- readRDS(file = here("data", "001_integrated-seurat_ADT.rds"))
#ElbowPlot(integrated.seurat_ADT, ndims = 100) # 50 PCs

#- Add PCA results to one Seurat object
integrated.seurat_GEX[["pca_ADT"]] <- CreateDimReducObject(embeddings = Embeddings(object = integrated.seurat_ADT, reduction = "pca"), assay = "integrated")

#- Run WNN
integrated.seurat_WNN <- FindMultiModalNeighbors(object = integrated.seurat_GEX,
                                                 reduction.list = list("pca", "pca_ADT"),
                                                 dims.list = list(1:50, 1:50),
                                                 modality.weight.name = c("RNA.weight", "ADT.weight"),
                                                 prune.SNN = 1/20)

#- UMAP
integrated.seurat_WNN <- RunUMAP(object = integrated.seurat_WNN, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_", seed.use = 1234, min.dist = .01, verbose = TRUE)

#- Clustering
set.seed(1234)
resolutions <- c(seq(from = .1, to = 1, by = .1), seq(from = 1.5, to = 2, by = .5))
for(i in resolutions) {
  integrated.seurat_WNN <- FindClusters(object = integrated.seurat_WNN, graph.name = "wsnn", algorithm = 3, resolution = i, n.start = 10, verbose = TRUE)
}

#- Output
saveRDS(object = integrated.seurat_WNN, file = here("data", "001_integrated-seurat_WNN.rds"))
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Processing - GEX + ADT = WNN
#- Open .rds
integrated.seurat_GEX <- readRDS(file = here("data", "001_integrated-seurat_GEX.rds"))
integrated.seurat_ADT <- readRDS(file = here("data", "001_integrated-seurat_ADT.rds"))
integrated.seurat_WNN <- readRDS(file = here("data", "001_integrated-seurat_WNN.rds"))
identical(colnames(integrated.seurat_WNN), colnames(integrated.seurat_GEX))
identical(colnames(integrated.seurat_WNN), colnames(integrated.seurat_ADT))

#- DietSeurat
DefaultAssay(integrated.seurat_WNN) <- "RNA"
reducedSize_integrated.seurat <- DietSeurat(object = integrated.seurat_WNN, 
                                            counts = TRUE, 
                                            data = TRUE, 
                                            assays = c("RNA", "ADT"), 
                                            dimreducs = NULL)

#- Add UMAPs
reducedSize_integrated.seurat[["umap_GEX"]] <- CreateDimReducObject(embeddings = Embeddings(object = integrated.seurat_GEX, reduction = "umap"), assay = "RNA")
reducedSize_integrated.seurat[["umap_ADT"]] <- CreateDimReducObject(embeddings = Embeddings(object = integrated.seurat_ADT, reduction = "umap"), assay = "ADT")
reducedSize_integrated.seurat[["umap_WNN"]] <- CreateDimReducObject(embeddings = Embeddings(object = integrated.seurat_WNN, reduction = "wnn.umap"), assay = "RNA")

#- Add meta.data
colnames(reducedSize_integrated.seurat@meta.data)
reducedSize_integrated.seurat@meta.data <- reducedSize_integrated.seurat@meta.data[, c(1:3, 31, 4:17, 33:34, 20, 27, 35:36, 37, 39:49)]

#- Preprocessing meta.data & ADT colnames
table(reducedSize_integrated.seurat@meta.data$sort_population)
reducedSize_integrated.seurat@meta.data$sort_population <- plyr::mapvalues(x = reducedSize_integrated.seurat@meta.data$sort_population, from = c("AgSpec", "AgSpec NOT"), to = c("AgSpec", "AgSpecNot"))
table(reducedSize_integrated.seurat@meta.data$sort_population)

#- Output
saveRDS(object = reducedSize_integrated.seurat, file = here("data", "001_integrated-seurat.rds"))
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
###--- Open .rds
integrated.seurat <- readRDS(file = here("data", "001_integrated-seurat.rds"))
# saveRDS(object = integrated.seurat, file = here("data", "001_integrated-seurat.rds"))
```

## UMAP - Technical factors

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=12, fig.height=10}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = nCount_RNA), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "# UMIs\nGEX") +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14)) -> plotA
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = nFeature_RNA), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "# genes") +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14)) -> plotB
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = percent.mt), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "% mito") +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14)) -> plotC
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = nCount_ADT), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "# UMIs\nADT") +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14)) -> plotD
cowplot::plot_grid(plotA, plotB, plotC, plotD, labels = "AUTO", align = c("hv"), ncol = 2)  
```

## UMAP - Batch  

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=7}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = batch), size = .01) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 8),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data)
for(j in data.ggplot$batch %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, batch == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: Batch", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-batch_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - PTID

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=7}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = ptid %>% as.character), size = .01) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 6),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data) 
for(j in data.ggplot$ptid %>% as.character() %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, ptid == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: PTID", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-ptid_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - Sorted Population

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=4}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = sort_population), size = .01) +
    scale_color_brewer(palette = "Set") +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ sort_population, ncol = 2) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data) 
for(j in data.ggplot$sort_population %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, sort_population == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: Sorted Population", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-sorted-population_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - Arm

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=4}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = arm), size = .01) +
    scale_color_brewer(palette = "Dark2") +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ arm, ncol = 2) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data) 
for(j in data.ggplot$arm %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, arm == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: Arm", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-arm_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - Stimulation

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=4}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = stim), size = .01) +
    scale_color_brewer(palette = "Set3", direction = -1) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ stim, ncol = 2) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data) 
for(j in data.ggplot$stim %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, stim == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: Stimulation", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-stimulation_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - Visit #

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=4}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = visit), size = .01) +
    scale_color_brewer(palette = "Paired") +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ visit, ncol = 2) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data) 
for(j in data.ggplot$visit %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, visit == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Summary: Visit", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-visitno_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - CD4 & CD8 (ADT)

Two ADTs (CD4 and CD8) are used to determine CD4+ and CD8+ T cells (using the `SCAMP` R package).  

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=6}
###--- UMAP
dt.SCAMP <- readRDS(file = here("output", "SCAMP", "dt_SCAMP.rds"))
dt.ggplot <- data.table(cell.id = rownames(integrated.seurat@meta.data), Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data)
dt.ggplot %>%
  filter(T_phenotype %in% c("CD4+", "CD8+")) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(x = wnnUMAP_1, y = wnnUMAP_2), size = .01, color = "grey70", data = dt.ggplot %>% filter(!(T_phenotype %in% c("CD4+", "CD8+")))) +
    geom_point(aes(color = T_phenotype), size = .01) +
    scale_color_manual(values = c("royalblue", "darkorange1")) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```

## UMAP - Markers

**Markers - GEX**

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=20, fig.height=12}
###--- UMAP
dt.mat <- integrated.seurat@assays$RNA@data[c("CD4", "CD8A", "CCR7", "GZMA", "TRAV1-2", "TRDV2", "MS4A1", "FCER1A"), ] %>% as.matrix() %>% t() %>% as.matrix()
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           dt.mat) %>%
  reshape2::melt(id.vars = c("wnnUMAP_1", "wnnUMAP_2")) %>%
  arrange(value) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = value), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "GEX") +
    facet_wrap(~ variable, ncol = 4) +
    theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 18),
            strip.text = element_text(size = 30),
            panel.grid = element_blank(),
            legend.text = element_text(size = 12),
            legend.position = "bottom")
```

**Markers - ADT**

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=20, fig.height=12}
###--- UMAP
dt.mat <- integrated.seurat@assays$ADT@data[c("CD4.1", "CD8a", "CD45RA", "CD45RO", "TCRVa7.2", "TCRVdelta2", "CD20", "CD11b"), ] %>% as.matrix() %>% t() %>% as.matrix()
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           dt.mat) %>%
  reshape2::melt(id.vars = c("wnnUMAP_1", "wnnUMAP_2")) %>%
  arrange(value) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = value), size = .01) +
    scale_color_viridis_c(option = "C") +
    labs(x = "UMAP-1", y = "UMAP-2", color = "ADT") +
    facet_wrap(~ variable, ncol = 4) +
    theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 18),
            strip.text = element_text(size = 30),
            panel.grid = element_blank(),
            legend.text = element_text(size = 12),
            legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
                          integrated.seurat@assays$ADT@data %>% as.matrix() %>% t() %>% as.matrix()) %>%
  reshape2::melt(id.vars = c("wnnUMAP_1", "wnnUMAP_2")) 
for(j in data.ggplot$variable %>% unique()){
  data.ggplot %>%
    filter(variable == j) %>%
    mutate(value = ifelse(value > 3, 3, value)) %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(aes(color = value), size = .01) +
      scale_color_gradient(low = "lightgrey", high = "darkgreen", breaks = c(0, 1, 2, 3), labels = c("0", "1", "2", ">3")) +
      labs(x = "UMAP-1", y = "UMAP-2", color = "ADT", title = paste("Summary: ADT", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 18),
            strip.text = element_text(size = 18),
            panel.grid = element_blank(),
            legend.text = element_text(size = 12),
            legend.position = "bottom") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-ADT-marker_", j, ".png")), width = 500, height = 600, res = 120)
  print(plot)
  dev.off()
}
```

## UMAP - Multimodal Reference Mapping

In Huo et al (2021), they have demonstrated how to use reference-mapping approach to annotate cell labels in a query dataset. This function is now implemented within `Seurat`. They have substantially improved the speed and memory requirements for integrative tasks including reference mapping, and also include new functionality to project query cells onto a previously computed UMAP visualization.  

```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Reference Mapping
#- Multimodal PBMC Reference Dataset
# reference <- LoadH5Seurat(file = "/fh/fast/mcelrath_j/vvoillet_working/CHIL/Projects/HVTN-602/CITE-seq/602_assay_CD3-/misc-docs/reference_pbmc_multimodal.h5seurat")
reference <- LoadH5Seurat(file = "~/OneDrive - Fred Hutchinson Cancer Research Center/CHIL/Projects/HVTN-602/CITE-seq/602_assay_CD3-/misc-docs/reference_pbmc_multimodal.h5seurat")
DimPlot(object = reference, reduction = "wnn.umap", group.by = "celltype.l2", label = TRUE, label.size = 3, repel = TRUE) + NoLegend()

#- SCTransform
# The reference was normalized using SCTransform(), so we use the same approach to normalize the query here.
# As it it best to remove batch effects & the MapQuery function cannot be used with integrated Seurat object,
# GEX normalization and mapping to reference will be done for each batch & pool individually, then results will be merged after mapping.
query.list <- foreach(i = integrated.seurat@meta.data$orig.ident %>% unique()) %do% {
  query <- subset(integrated.seurat, orig.ident == i)
  DefaultAssay(query) <- "RNA"
  query <- SCTransform(query, verbose = TRUE)
  query %>% return()
}

#- Anchors
# Find anchors between reference and query
anchors.list <- foreach(i = 1:length(query.list)) %do% {
  cat(i)
  anchors <- FindTransferAnchors(reference = reference,
                                 query = query.list[[i]],
                                 normalization.method = "SCT",
                                 reference.reduction = "spca",
                                 dims = 1:50, 
                                 verbose = TRUE)
  anchors %>% return()
}

#- Prediction
registerDoMC(2)
prediction.list <- foreach(i = 1:length(anchors.list)) %dopar% {
  cat(i)
  prediction <- MapQuery(anchorset = anchors.list[[i]],
                         query = query.list[[i]],
                         reference = reference,
                         refdata = list(celltype.l1 = "celltype.l1",
                                        celltype.l2 = "celltype.l2"),
                          reference.reduction = "spca", 
                  reduction.model = "wnn.umap", 
                  verbose = TRUE)
  prediction %>% return()
}

#- meta.data
dt.prediction <- foreach(i = 1:length(prediction.list)) %do% {
  dt.metadata <- prediction.list[[i]]@meta.data
  dt.metadata %>% return()
} %>% bind_rows()
length(intersect(rownames(dt.prediction), rownames(integrated.seurat@meta.data)))
dt.prediction <- dt.prediction[rownames(integrated.seurat@meta.data), ]
identical(rownames(dt.prediction), rownames(integrated.seurat@meta.data)) # TRUE
integrated.seurat@meta.data$predicted.celltype.l1.score <- dt.prediction$predicted.celltype.l1.score
integrated.seurat@meta.data$predicted.celltype.l1 <- dt.prediction$predicted.celltype.l1
integrated.seurat@meta.data$predicted.celltype.l2.score <- dt.prediction$predicted.celltype.l2.score
integrated.seurat@meta.data$predicted.celltype.l2 <- dt.prediction$predicted.celltype.l2
head(integrated.seurat@meta.data)

#- Output
saveRDS(object = dt.prediction, file = here("output", "001_output", "Seurat_prediction.rds"))
saveRDS(object = integrated.seurat, file = here("data", "001_integrated-seurat.rds"))
```
```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=7}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = predicted.celltype.l1), size = .01) +
    scale_color_brewer(palette = "Paired") +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    guides(colour = guide_legend(override.aes = list(size = 4, alpha = 1))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 10),
          legend.position = "bottom")
```

With the `Seurat` prediction, we can go into details - see in 001_output density plots.

```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE, fig.width=6, fig.height=8}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = predicted.celltype.l2), size = .01) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    guides(colour = guide_legend(override.aes = list(size = 4, alpha = 1))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 8),
          legend.position = "bottom")
```
```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"), integrated.seurat@meta.data)
for(j in data.ggplot$predicted.celltype.l2 %>% unique()){
  data.ggplot %>%
    ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
      geom_point(color = "grey80", size = .01) +
      ggpointdensity::geom_pointdensity(aes(x = wnnUMAP_1, y = wnnUMAP_2), data = subset(data.ggplot, predicted.celltype.l2 == j), size = .01) +
      viridis::scale_color_viridis() +
      labs(x = "UMAP-1", y = "UMAP-2", color = NULL, title = paste("Prediction:", j)) +
      theme_bw() +
      theme(axis.text = element_blank(),
            axis.ticks = element_blank(),
            axis.title = element_text(size = 14),
            panel.grid = element_blank(),
            legend.text = element_text(size = 6),
            legend.position = "none") -> plot
  png(here::here("output", "001_output", paste0("WNN_UMAP-by-seurat-prediction_", j, ".png")), width = 500, height = 550, res = 120)
  print(plot)
  dev.off()
}
```

## Clustering

With a `resolution = 1.5`, we have 35 clusters    

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=4, fig.height=4}
###--- UMAP
Idents(integrated.seurat) <- factor(x = integrated.seurat@meta.data$wsnn_res.1.5, levels = 0:34) 
data.table(Embeddings(object = integrated.seurat, "umap_WNN"), wsnn_res.1.5 = Idents(integrated.seurat)) %>%
  select(wnnUMAP_1, wnnUMAP_2, wsnn_res.1.5) %>%
  reshape2::melt(id.vars = c("wnnUMAP_1", "wnnUMAP_2")) %>%
  mutate(resolution = str_replace(string = variable, pattern = "wsnn_res.", replacement = "resolution ")) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = value), size = .01) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ resolution, ncol = 5) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "none")
```

When aiming to identify markers of cell state independent of experimental condition, we include batch, donor, arm, stimulation and visit as covariates to the model. To identify gene markers, the `FindAllMarkers` R function is used with `test.use = MAST`, `latent.vars = c("nCount_RNA", "batch", "ptid", "arm", "stimulation", "visit")`, and `max.cells.per.ident = 500`.  
To identify ADT markers, the `FindAllMarkers` R function is used with `test.use = LR`, `latent.vars = c("batch", "ptid", "arm", "stimulation", "visit")`, and `max.cells.per.ident = 500`. We merge clusters that do not exhibit clear evidence of separation.  

```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- UMAP
data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
           integrated.seurat@meta.data) %>%
  select(wnnUMAP_1, wnnUMAP_2, wsnn_res.1.5) %>%
  mutate(wsnn_res.1.5 = factor(x = wsnn_res.1.5, levels = 0:34)) %>%
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = wsnn_res.1.5), size = .01) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    facet_wrap(~ wsnn_res.1.5, ncol = 5) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "none")
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Cluster annotation - GEX & ADT
#- Variables
annotations <- read.table(here("misc-docs", "2020_gene-annotation.txt"), header = TRUE, sep = ",")
Idents(integrated.seurat) <- factor(x = integrated.seurat@meta.data$wsnn_res.1.5, levels = 0:34) 
Idents(integrated.seurat) %>% table()
integrated.seurat@meta.data$batch %>% class()
integrated.seurat@meta.data$ptid %>% class()
integrated.seurat@meta.data$ptid <- integrated.seurat@meta.data$ptid %>% as.character()
integrated.seurat@meta.data$arm %>% class()
integrated.seurat@meta.data$stim %>% class()
integrated.seurat@meta.data$visit %>% class()

#- MAST (GEX)
# Perform differential expression for each of the merged clusters
DefaultAssay(integrated.seurat) <- "RNA"
markers.clusters_GEX <- FindAllMarkers(object = integrated.seurat, 
                                       max.cells.per.ident = 500,
                                       only.pos = TRUE, 
                                       test.use = "MAST", 
                                       latent.vars = c("nCount_RNA", "batch", "ptid", "arm", "stim", "visit"),
                                       verbose = TRUE)
markers.clusters_GEX <- markers.clusters_GEX %>%
  filter(p_val_adj < 1e-05) %>%
  arrange(cluster, p_val_adj, avg_log2FC) %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]), by = c("gene" = "gene_name")) 
markers.clusters_GEX %>%
  write_csv(file = here("output", "001_output", "GEX_markers-clusters_v1.csv"))

#- Top 20 (GEX)
markers.clusters_GEX %>%
  group_by(cluster) %>% 
  top_n(n = 20, wt = avg_log2FC) %>%
  View("Top genes")

#- LR (ADT)
# Perform differential expression for each of the merged clusters
DefaultAssay(integrated.seurat) <- "ADT"
markers.clusters_ADT <- FindAllMarkers(object = integrated.seurat, 
                                       max.cells.per.ident = 500,
                                       only.pos = TRUE, 
                                       test.use = "LR", 
                                       latent.vars = c("batch", "ptid", "arm", "stim", "visit"),
                                       verbose = TRUE)
markers.clusters_ADT <- markers.clusters_ADT %>%
  filter(p_val_adj < 1e-05) %>%
  arrange(cluster, p_val_adj, avg_log2FC)
markers.clusters_ADT %>%
  write_csv(file = here("output", "001_output", "ADT_markers-clusters_v1.csv"))

#- Top 20 (ADT)
markers.clusters_ADT %>%
  group_by(cluster) %>% 
  top_n(n = 20, wt = avg_log2FC) %>%
  View("Top ADTs")

#- Ridge plots
RidgePlot(object = integrated.seurat, features = c("TCRVa7.2", "CD103", "TCRVdelta2", "TCRva24-Ja18", "CD183", "CD1c"), group.by = "wsnn_res.1.5")
RidgePlot(object = integrated.seurat, features = c("CD4.1", "CD8", "CD45RA"), group.by = "wsnn_res.1.5")

#- Seurat predictions
table(Idents(integrated.seurat), integrated.seurat@meta.data$predicted.celltype.l2)
table(Idents(integrated.seurat), integrated.seurat@meta.data$sort_population)

#- Proposed annotation
# 0 - CCL5+ Activated CD8+ T
# 1 - Naive CD4+ T 
# 2 - Naive CD4+ T 
# 3 - EM CD4+ T (AgSpec) 
# 4 - Naive CD4+ T
# 5 - Early Differentiated EM CD4+ T 
# 6 - Naive CD8+ T
# 7 - EM CD4+ T
# 8 - Naive CD8+ T 
# 9 - CCL5+ Activated CD8+ T 
# 10 - Early Differentiated EM CD4+ T 
# 11 - CCL5+GZMK+ Activated CD8+ T
# 12 - EM CD4+ T (AgSpec) 
# 13 - Regulatory T
# 14 - Activated CD8+ T (AgSpec) 
# 15 - MAIT
# 16 - Early Differentiated EM CD4+ T (AgSpec) 
# 17 - CD103+ Activated CD8+ T
# 18 - MAIT 
# 19 - ADT-high 
# 20 - Naive CD4+ T 
# 21 - CXCR3+ Activated CD8+ T 
# 22 - Naive CD4+ T 
# 23 - HSP-genes 
# 24 - IFI27+ T 
# 25 - Naive CD4+ T
# 26 - EM CD4+ T (AgSpec) 
# 27 - KLRB1+ T 
# 28 - EM CD4+ T (AgSpec) 
# 29 - CD103+ EM CD4+ T
# 30 - B
# 31 - EM CD4+ T (AgSpec) 
# 32 - Naive CD4+ T
# 33 - gdT
# 34 - DCs/Monocytes

#- Add results to meta.data
integrated.seurat@meta.data$integrated.clusters <- integrated.seurat@meta.data$wsnn_res.1.5
integrated.seurat@meta.data$integrated.annotation.clusters <- plyr::mapvalues(x = integrated.seurat@meta.data$integrated.clusters, 
                                                                              from = 0:34, 
                                                                              to = c("CCL5+ Activated CD8+ T",
                                                                                     "Naive CD4+ T", 
                                                                                     "Naive CD4+ T", 
                                                                                     "EM CD4+ T (AgSpec)", 
                                                                                     "Naive CD4+ T", 
                                                                                     "Early Differentiated EM CD4+ T", 
                                                                                     "Naive CD8+ T", 
                                                                                     "EM CD4+ T", 
                                                                                     "Naive CD8+ T", 
                                                                                     "CCL5+ Activated CD8+ T", 
                                                                                     "Early Differentiated EM CD4+ T", 
                                                                                     "CCL5+GZMK+ Activated CD8+ T", 
                                                                                     "EM CD4+ T (AgSpec)", 
                                                                                     "Regulatory T", 
                                                                                     "Activated CD8+ T (AgSpec)", 
                                                                                     "MAIT", 
                                                                                     "Early Differentiated EM CD4+ T (AgSpec)", 
                                                                                     "CD103+ Activated CD8+ T", 
                                                                                     "MAIT", 
                                                                                     "ADT-high", 
                                                                                     "Naive CD4+ T", 
                                                                                     "CXCR3+ Activated CD8+ T", 
                                                                                     "Naive CD4+ T", 
                                                                                     "HSP-genes", 
                                                                                     "IFI27+ T", 
                                                                                     "Naive CD4+ T", 
                                                                                     "EM CD4+ T (AgSpec)", 
                                                                                     "KLRB1+ T", 
                                                                                     "EM CD4+ T (AgSpec)", 
                                                                                     "CD103+ EM CD4+ T", 
                                                                                     "B", 
                                                                                     "EM CD4+ T (AgSpec)", 
                                                                                     "Naive CD4+ T", 
                                                                                     "gdT", 
                                                                                     "DCs/Monocytes")) %>% as.character
integrated.seurat@meta.data$integrated.clusters <- plyr::mapvalues(x = integrated.seurat@meta.data$integrated.annotation.clusters,
                                                                   from = c("Naive CD4+ T", 
                                                                            "Early Differentiated EM CD4+ T", 
                                                                            "Early Differentiated EM CD4+ T (AgSpec)", 
                                                                            "EM CD4+ T", 
                                                                            "EM CD4+ T (AgSpec)",
                                                                            "CD103+ EM CD4+ T",
                                                                            "Naive CD8+ T", 
                                                                            "CCL5+ Activated CD8+ T",
                                                                            "CCL5+GZMK+ Activated CD8+ T",
                                                                            "CD103+ Activated CD8+ T",
                                                                            "CXCR3+ Activated CD8+ T",
                                                                            "Activated CD8+ T (AgSpec)",
                                                                            "Regulatory T",
                                                                            "MAIT",
                                                                            "gdT",
                                                                            "KLRB1+ T",
                                                                            "IFI27+ T",
                                                                            "HSP-genes", 
                                                                            "ADT-high",
                                                                            "B",
                                                                            "DCs/Monocytes"),
                                                                   to = paste("Seurat", 1:21))
integrated.seurat@meta.data$integrated.clusters <- factor(x = integrated.seurat@meta.data$integrated.clusters, levels = paste("Seurat", 1:21))
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Cluster annotation - GEX & ADT - version 2 (merged clusters)
Idents(integrated.seurat) <- integrated.seurat@meta.data$integrated.clusters
Idents(integrated.seurat) %>% table()

#- MAST (GEX)
# Perform differential expression for each of the merged clusters
DefaultAssay(integrated.seurat) <- "RNA"
markers.clusters_GEX_2 <- FindAllMarkers(object = integrated.seurat, 
                                         max.cells.per.ident = 500,
                                         only.pos = TRUE, 
                                         test.use = "MAST", 
                                         latent.vars = c("nCount_RNA", "batch", "ptid", "arm", "stim", "visit"),
                                         verbose = TRUE)
markers.clusters_GEX_2 <- markers.clusters_GEX_2 %>%
  filter(p_val_adj < 1e-05) %>%
  arrange(cluster, p_val_adj, avg_log2FC) %>%
  left_join(y = unique(annotations[, c("gene_name", "description")]), by = c("gene" = "gene_name")) %>%
  filter(!str_detect(string = gene, pattern = "RP")) %>%
  mutate(`Annotation - Valentin` = plyr::mapvalues(x = cluster, 
                                                   to = c("Naive CD4+ T", 
                                                          "Early Differentiated EM CD4+ T", 
                                                          "Early Differentiated EM CD4+ T (AgSpec)", 
                                                          "EM CD4+ T", 
                                                          "EM CD4+ T (AgSpec)",
                                                          "CD103+ EM CD4+ T",
                                                          "Naive CD8+ T", 
                                                          "CCL5+ Activated CD8+ T",
                                                          "CCL5+GZMK+ Activated CD8+ T",
                                                          "CD103+ Activated CD8+ T",
                                                          "CXCR3+ Activated CD8+ T",
                                                          "Activated CD8+ T (AgSpec)",
                                                          "Regulatory T",
                                                          "MAIT",
                                                          "gdT",
                                                          "KLRB1+ T",
                                                          "IFI27+ T",
                                                          "HSP-genes", 
                                                          "ADT-high",
                                                          "B",
                                                          "DCs/Monocytes"), 
                                                   from = paste("Seurat", 1:21))) 
markers.clusters_GEX_2 %>%
  write_csv(file = here("output", "001_output", "GEX_markers-clusters_v2.csv"))

#- Top 20 (GEX)
markers.clusters_GEX_2 %>%
  group_by(cluster) %>% 
  filter(!str_detect(string = gene, pattern = "RP")) %>%
  top_n(n = 20, wt = avg_log2FC) %>%
  View("Top genes")

#- LR (ADT)
# Perform differential expression for each of the merged clusters
DefaultAssay(integrated.seurat) <- "ADT"
markers.clusters_ADT_2 <- FindAllMarkers(object = integrated.seurat, 
                                         max.cells.per.ident = 500,
                                         only.pos = TRUE, 
                                         test.use = "LR", 
                                         latent.vars = c("batch", "ptid", "arm", "stim", "visit"),
                                         verbose = TRUE)
markers.clusters_ADT_2 <- markers.clusters_ADT_2 %>%
  filter(p_val_adj < 1e-05) %>%
  arrange(cluster, p_val_adj, avg_log2FC) %>%
  mutate(`Annotation - Valentin` = plyr::mapvalues(x = cluster, 
                                                   to = c("Naive CD4+ T", 
                                                          "Early Differentiated EM CD4+ T", 
                                                          "Early Differentiated EM CD4+ T (AgSpec)", 
                                                          "EM CD4+ T", 
                                                          "EM CD4+ T (AgSpec)",
                                                          "CD103+ EM CD4+ T",
                                                          "Naive CD8+ T", 
                                                          "CCL5+ Activated CD8+ T",
                                                          "CCL5+GZMK+ Activated CD8+ T",
                                                          "CD103+ Activated CD8+ T",
                                                          "CXCR3+ Activated CD8+ T",
                                                          "Activated CD8+ T (AgSpec)",
                                                          "Regulatory T",
                                                          "MAIT",
                                                          "gdT",
                                                          "KLRB1+ T",
                                                          "IFI27+ T",
                                                          "HSP-genes", 
                                                          "ADT-high",
                                                          "B",
                                                          "DCs/Monocytes"), 
                                                   from = paste("Seurat", 1:21))) 
markers.clusters_ADT_2 %>%
  write_csv(file = here("output", "001_output", "ADT_markers-clusters_v2.csv"))

#- Top 20 (ADT)
markers.clusters_ADT_2 %>%
  group_by(cluster) %>% 
  top_n(n = 20, wt = avg_log2FC) %>%
  View("Top ADTs")
```

According to these markers and after merging clusters that do not exhibit clear evidence of separation, we have  

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=11}
###--- UMAP
#- Data
dt.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_WNN"),
                        integrated.seurat@meta.data)

#- Annotation
centers <- dt.ggplot %>% group_by(integrated.clusters, integrated.annotation.clusters) %>% summarize(UMAP_1 = median(wnnUMAP_1), UMAP_2 = median(wnnUMAP_2))

#- Plot
my.colors <- c(RColorBrewer::brewer.pal(n = 12, name = "Paired"), RColorBrewer::brewer.pal(n = 8, name = "Dark2"), "black")
names(my.colors) <- paste("Seurat", 1:21)
dt.ggplot %>%  
  ggplot(aes(x = wnnUMAP_1, y = wnnUMAP_2)) +
    geom_point(aes(color = integrated.clusters), size = .01, alpha = .2) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    scale_color_manual(values = my.colors) +
    geom_text(aes(x = UMAP_1, y = UMAP_2, label = integrated.annotation.clusters), size = 3, color = "black", data = centers, fontface = "bold") +
    guides(colour = guide_legend(override.aes = list(size = 4, alpha = 1))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "bottom")
```

UMAP coordinates using GEX, we have    

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=6}
###--- UMAP (GEX)
#- Data
dt.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_GEX"),
                        integrated.seurat@meta.data)

#- Plot
my.colors <- c(RColorBrewer::brewer.pal(n = 12, name = "Paired"), RColorBrewer::brewer.pal(n = 8, name = "Dark2"), "black")
names(my.colors) <- paste("Seurat", 1:21)
dt.ggplot %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2)) +
    geom_point(aes(color = integrated.clusters), size = .01, alpha = .2) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    scale_color_manual(values = my.colors) +
    guides(colour = guide_legend(override.aes = list(size = 4, alpha = 1))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "none") # png 1000 x 1100
```

UMAP coordinates using ADT, we have    

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=6}
###--- UMAP (ADT)
#- Data
dt.ggplot <- data.table(Embeddings(object = integrated.seurat, "umap_ADT"),
                        integrated.seurat@meta.data)

#- Plot
my.colors <- c(RColorBrewer::brewer.pal(n = 12, name = "Paired"), RColorBrewer::brewer.pal(n = 8, name = "Dark2"), "black")
names(my.colors) <- paste("Seurat", 1:21)
dt.ggplot %>%  
  ggplot(aes(x = umap_adt_1, y = umap_adt_2)) +
    geom_point(aes(color = integrated.clusters), size = .01, alpha = .2) +
    labs(x = "UMAP-1", y = "UMAP-2", color = NULL) +
    scale_color_manual(values = my.colors) +
    guides(colour = guide_legend(override.aes = list(size = 4, alpha = 1))) +
    theme_bw() +
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_text(size = 14),
          panel.grid = element_blank(),
          legend.text = element_text(size = 12),
          legend.position = "none") # png 1000 x 1100
```

**Heatmap - GEX**

Not shown.  

```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- Heatmap
#- Top 10
DEGs <- markers.clusters_GEX_2 %>%
  group_by(cluster) %>% 
  filter(!str_detect(string = gene, pattern = "RP")) %>%
  top_n(n = 10, wt = avg_log2FC) %>%
  pull(gene)

#- Heatmap (downsample - Seurat)
my.colors <- c(RColorBrewer::brewer.pal(n = 12, name = "Paired"), RColorBrewer::brewer.pal(n = 8, name = "Dark2"), "black")
names(my.colors) <- paste("Seurat", 1:21)

DefaultAssay(integrated.seurat) <- "RNA"
Idents(integrated.seurat) <- factor(x = integrated.seurat@meta.data$integrated.clusters, levels = paste("Seurat", 1:21)) 
seurat.tmp <- subset(integrated.seurat, downsample = 100)
seurat.tmp <- ScaleData(object = seurat.tmp, features = DEGs, verbose = TRUE)

png(filename = here("output", "001_output", "GEX_heatmap-top10.png"), width = 1000, height = 1300, bg = "white")
DoHeatmap(seurat.tmp, features = DEGs, group.colors = my.colors)
dev.off()
```

**Ridge plots - ADT**

Not shown.  

```{r echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
###--- RidgePlot - ADTs
my.colors <- c(RColorBrewer::brewer.pal(n = 12, name = "Paired"), RColorBrewer::brewer.pal(n = 8, name = "Dark2"), "black")
names(my.colors) <- paste("Seurat", 1:21)

pdf(file = here("output", "001_output", "ADT_ridgePlots.pdf"), width = 5, height = 8, bg = "white", pointsize = 12)
for(i in setdiff(rownames(integrated.seurat@assays$ADT@data), "CD4.2")){
  data.table(integrated.seurat@meta.data, ADT = integrated.seurat@assays$ADT@data[i, ]) %>%
    mutate(integrated.clusters = factor(x = integrated.clusters, levels = paste("Seurat", 1:21))) %>%
    ggplot(aes(y = integrated.clusters, x = ADT, fill = integrated.clusters)) +
      ggridges::geom_density_ridges(bandwidth = .2) +
      geom_vline(xintercept = 0, color = "grey20", alpha = .7, lty = 2) +
      labs(x = "Normalized ADT expression", title = paste(str_remove(string = i, pattern = "[.]1"), "(ADT)")) +
      scale_fill_manual(values = my.colors) +
      theme_bw() +
      theme(axis.text = element_text(size = 12),
            axis.title.x = element_text(size = 14, face = "plain"),
            axis.title.y = element_blank(),
            title = element_text(size = 16, face = "bold"),
            panel.grid = element_blank(),
            legend.text = element_text(size = 12),
            legend.position = "none") -> plot
  print(plot)
}
dev.off()
```

